spec/requests/api/v1/users_request_spec.rb

```diff
@@ -0,0 +1,133 @@
+
+      users = JSON.parse(response.body, symbolize_names: true)
+      expect(users[:data].count).to eq(5)
+      # require 'pry'; binding.pry
+      users[:data].each do |user|
+        expect(users).to be_a(Hash)
+        expect(users[:data]).to be_an(Array)
+        
+        expect(user).to have_key(:id)
+        expect(user[:id]).to be_an(String)
+        
+        expect(user).to have_key(:attributes)
+        
+        expect(user[:attributes]).to have_key(:name)
+        expect(user[:attributes][:name]).to be_a(String)
+
+        expect(user[:attributes]).to have_key(:email)
+        expect(user[:attributes][:email]).to be_a(String)
+      end
+    end
+  end
+
+  describe "User Show" do
+    it "can get one user by its id" do
+      id = create(:user).id
+
+      get api_v1_user_path(id)
+  
+      user = JSON.parse(response.body, symbolize_names: true)
+  
+      expect(response).to be_successful
+      
+      expect(user).to be_a(Hash)
+      
+      expect(user[:data]).to have_key(:id)
+      expect(user[:data][:id]).to be_an(String)
+  
+      expect(user[:data][:attributes]).to have_key(:name)
+      expect(user[:data][:attributes][:name]).to be_a(String)
+  
+      expect(user[:data][:attributes]).to have_key(:email)
+      expect(user[:data][:attributes][:email]).to be_a(String)
+    end
+  end
+
+  describe "User Create" do
+    it "can create a new user" do
+      user_params = ({
+                      name: 'James Sullivan',
+                      email: 'james.p.sullivan@aol.com',
+                      password: "gw45635yhethet5"
+                    })
+      headers = {"CONTENT_TYPE" => "application/json"}
+    
+      post api_v1_users_path, headers: headers, params: JSON.generate(user: user_params)
+      
+      created_user = User.last
+
+      expect(response).to be_successful
+      expect(created_user.name).to eq(user_params[:name])
+      expect(created_user.email).to eq(user_params[:email])
+      expect(created_user.password).to eq(user_params[:password_digest])
+    end
+
+    it "sad path; will send an error if user is not created" do 
+      user_params = ({
+        name: "",
+        email: 'james.p.sullivan@aol.com',
+        role: 1,
+        password: "gw45635yhethet5"
+      })
+      headers = {"CONTENT_TYPE" => "application/json"}
+  
+      post api_v1_users_path, headers: headers, params: JSON.generate(user: user_params)
+
+      expect(response).to_not be_successful 
+      expect(response).to have_http_status(401)
+    end
+  end
+
+  describe "User Update" do
+    it "can update an existing user" do
+      id = create(:user).id
+      previous_name = User.last.name
+      user_param = { name: "P. Sherman" }
+      headers = {"CONTENT_TYPE" => "application/json"}
+    
+      patch api_v1_user_path(id), headers: headers, params: JSON.generate({user: user_param})
+  
+      user = User.find_by(id: id)
+      expect(response).to be_successful
+      expect(user.name).to_not eq(previous_name)
```

Comment:
Lmk if this is doesn't really have value because this is already a good test, but should we check to see that the user.email == previous_user.email? This is just checking to make sure that said 'user' we are updating, only the desired attribute is being changed?

Reply:
I just want to clarify, are you suggesting I add testing to show that other attributes can be updated as well? My intention with testing the name change alone is to show that you don't need all attributes updated in order to make the update, but I can easily create another test for testing that an email can be updated.

Reply:
Clarifying: It is to see that nothing else changed expect the desired attribute, the name. Really what I'm wanting to show is that the email is the same after the update. 

Example,

user1 = { name: "bob", email: "bob@gmail.com", password: "shhhhhh" }
id = user1.id
previous_user = User.last
user_param = { name: "Bobby" }
headers = {"CONTENT_TYPE" => "application/json"}

patch api_v1_user_path(id), headers: headers, params: JSON.generate({user: user_param})

user = User.find_by(id: id)
expect(response).to be_successful
expect(user1.name).to_not eq(previous_user.name)
expect(user1.name).to eq("Bobby")
expect(user1.email).to eq("bob@gmail.com")

Reply:
I understand what you are saying with this Eric. Very thorough. I think we can merge now and add more to our tests if we see fit in the refactor!